internal/command/cmd.go

```diff
@@ -36,67 +37,89 @@ const (
+// RetryCommandWithOutput will retry the given command,  with exponential backoff, maxing at 5 min, for numOfRetries times
+// commandTimeout is used to terminate the command after specified time, regardless if it was successful or not
+// this prevents commands to be executing indefinitely
 // returns (nil, error) if all retries fail, (output, nil) otherwise
-func (c Cmd) RetryCommandWithOutput(numOfRetries int) ([]byte, error) {
-	// have an initial backoff before trying again.
-	log.Info().Msgf("Next retry in %ds...", 15)
-	time.Sleep(15 * time.Second)
-
+func (c *Cmd) RetryCommandWithOutput(numOfRetries, commandTimeout int) ([]byte, error) {
 	var err error
+	var out []byte
 	for i := 1; i <= numOfRetries; i++ {
-		log.Warn().Msgf("Retrying command %s... (%d/%d)", c.Command, i, numOfRetries)
-		cmd := exec.Command("bash", "-c", c.Command)
-		cmd.Dir = c.Dir
-		cmd.Stdout = c.Stdout
-		cmd.Stderr = c.Stderr
-
-		var out []byte
-		out, err = cmd.CombinedOutput()
-		if err == nil {
+		backoff := getNewBackoff(i)
+		log.Info().Msgf("Next retry in %ds...", backoff)
+		time.Sleep(time.Duration(backoff) * time.Second)
+
+		if out, err = c.executeWithOutput(i, numOfRetries, commandTimeout); err == nil {
 			log.Info().Msgf("The %s was successful after %d retry", c.Command, i)
 			return out, nil
 		}
-
 		log.Warn().Msgf("Error encountered while executing %s : %v", c.Command, err)
-		backoff := 10 * i
-		log.Info().Msgf("Next retry in %ds...", backoff)
-		time.Sleep(time.Duration(backoff) * time.Second)
 	}
 	log.Error().Msgf("Command %s was not successful after %d retries", c.Command, numOfRetries)
 	return nil, err
 }
 
+// execute executes the cmd with context canceled after commandTimeout seconds
+// returns error if unsuccessful, nil otherwise
+func (c *Cmd) execute(i, numOfRetries, commandTimeout int) error {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(commandTimeout)*time.Second)
+	defer cancel()
+	log.Warn().Msgf("Retrying command %s... (%d/%d)", c.Command, i, numOfRetries)
+	return c.buildCmd(ctx).Run()
+}
+
+// executeWithOutput executes the cmd with context canceled after commandTimeout seconds
+// returns error, nil if unsuccessful, nil, output otherwise
+func (c *Cmd) executeWithOutput(i, numOfRetries, commandTimeout int) ([]byte, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(commandTimeout)*time.Second)
+	defer cancel()
+	log.Warn().Msgf("Retrying command %s... (%d/%d)", c.Command, i, numOfRetries)
+	return c.buildCmd(ctx).CombinedOutput()
+}
+
+// buildCmd prepares a exec.Cmd datastructure with context
+func (c *Cmd) buildCmd(ctx context.Context) *exec.Cmd {
+	cmd := exec.CommandContext(ctx, "bash", "-c", c.Command)
+	cmd.Dir = c.Dir
+	cmd.Stdout = c.Stdout
+	cmd.Stderr = c.Stderr
+	return cmd
+}
+
+// getNewBackoff returns a new backoff 5 * (2 ^ iteration), with the hard limit set at maxBackoff
+func getNewBackoff(iteration int) int {
+	backoff := 5 * (2 ^ iteration)
```

Comment:
> 0 is not passed in anyway? 

As you said, 0 is fine since 2^0 = 1

>why are we iterating from 1 then? 

For easier logging

> in that case, we'd need to make sure that iteration >= 0

Don't necessarily have to, since the function works with negative numbers as well. i.e. `5 * 2^(-1) = 2.5`. The backoff function is clearly stated in the comment, thus if developer wants to start with a backoff smaller than 5, they can.